SCAutolib/models/gui.py

```diff
@@ -0,0 +1,281 @@
+        # So this effectively does nothing
+        keyboard.send('enter')
+
+
+class GUI():
+    def __init__(self):
+        # Create the directory for screenshots
+        # TODO parametrize?
+        self.screenshot_directory = '/tmp/SC-tests'
+        os.makedirs(self.screenshot_directory, exist_ok=True)
+
+        self.mouse = Mouse()
+
+        self.kb = KB()
+
+    def __enter__(self):
+        self.screen = Screen(self.screenshot_directory)
+        # By restarting gdm, the system gets into defined state
+        run(['systemctl', 'restart', 'gdm'], check=True)
+        # Cannot screenshot before gdm starts displaying
+        # This would break the display
+        sleep(5)
+
+        return self
+
+    def __exit__(self, type, value, traceback):
+        # Gather the screenshots, generate report
+        pass
+
+    @staticmethod
+    def _image_to_data(filename):
+        """
+        Convert screenshot into dataframe of words with their coordinates.
+        """
+        image = cv2.imread(filename)
+        grayscale = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
+        upscaled = cv2.resize(grayscale,
+                              dsize=None, fx=2, fy=2,
+                              interpolation=cv2.INTER_LANCZOS4)
+        _, binary = cv2.threshold(upscaled, 120, 255, cv2.THRESH_BINARY_INV)
+        image_data_str = pytesseract.image_to_data(binary)
+        df = pd.read_csv(StringIO(image_data_str),
+                         sep='\t', lineterminator='\n')
+        return df
+
+    @staticmethod
+    def _images_same(filename1: str, filename2: str):
+        """Compare two images, return True if they are completely identical."""
+        im1 = cv2.imread(filename1)
+        im2 = cv2.imread(filename2)
+
+        # Is the resolution the same
+        if im1.shape != im2.shape:
+            return False
+
+        # Check if value of every pixel is the same
+        if np.bitwise_xor(im1, im2).any():
+            return False
+
+        return True
+
+    def click_on(self, key: str, timeout: float = 30):
+        end_time = time() + timeout
+
+        # Repeat screenshotting, until the key is found
+        while time() < end_time:
+            # Capture the screenshot
+            screenshot = self.screen.screenshot()
+            df = self._image_to_data(screenshot)
+            selection = df['text'] == key
+
+            # If there is no matching word, try again
+            if selection.sum() == 0:
+                continue
+
+            # Exactly one word matching, exit the loop
+            elif selection.sum() == 1:
+                item = df[selection]
+                break
+
+            # More than one word matches, choose the first match
+            # Probably deterministic, but it should not be relied upon
+            else:
+                print('Too many to choose from')
+                # TODO replace the print function
+                item = df.iloc[0]
+
+        if time() >= end_time:
```

Comment:
It seems to me easy to make such condition "false positive" with changes in while statement above. It may easily happen more code will be added leading to false positives. This check should rather be based on content if 'item'.